velox/functions/prestosql/types/JsonType.cpp

```diff
@@ -62,57 +67,355 @@ void generateJsonTyped(
+    if (inputVector->isNullAt(row)) {
+      if constexpr (throwIfNull) {
+        VELOX_FAIL("Unexpected null value in input vector.");
+      } else {
+        flatResult.set(row, "null");
+      }
+    } else {
+      result.clear();
+      generateJsonTyped(*inputVector, row, result);
+
+      flatResult.set(row, StringView{result});
+    }
+  });
+}
+
+// Forward declaration.
+void castToJsonFromArray(
+    const BaseVector& input,
+    exec::EvalCtx* context,
+    const SelectivityVector& rows,
+    FlatVector<StringView>& flatResult);
+
+void castToJsonFromMap(
+    const BaseVector& input,
+    exec::EvalCtx* context,
+    const SelectivityVector& rows,
+    FlatVector<StringView>& flatResult);
+
+void castToJsonFromRow(
+    const BaseVector& input,
+    exec::EvalCtx* context,
+    const SelectivityVector& rows,
+    FlatVector<StringView>& flatResult);
+
+// Casts complex-type input vectors to Json type.
+template <
+    TypeKind kind,
+    typename std::enable_if<!TypeTraits<kind>::isPrimitiveType, int>::type = 0>
+void castToJson(
+    const BaseVector& input,
+    exec::EvalCtx* context,
+    const SelectivityVector& rows,
+    FlatVector<StringView>& flatResult) {
+  if constexpr (kind == TypeKind::ARRAY) {
+    castToJsonFromArray(input, context, rows, flatResult);
+  } else if constexpr (kind == TypeKind::MAP) {
+    castToJsonFromMap(input, context, rows, flatResult);
+  } else if constexpr (kind == TypeKind::ROW) {
+    castToJsonFromRow(input, context, rows, flatResult);
+  } else {
+    VELOX_NYI(
+        "Casting from {} to JSON is not supported yet.",
+        input.type()->toString());
+  }
+}
+
+// Calculate the sum of Json string sizes of all elements within the range
+// [offset, offset + size) in vector wrapped with indices.
+FOLLY_ALWAYS_INLINE size_t getSumOfElementsStringSize(
+    const FlatVector<StringView>& vector,
+    const vector_size_t* indices,
+    vector_size_t offset,
+    vector_size_t size) {
+  size_t sum = 0;
+  for (auto i = offset; i < offset + size; ++i) {
+    sum += vector.valueAt(indices[i]).size();
+  }
+
+  return sum;
+}
+
+void castToJsonFromArray(
+    const BaseVector& input,
+    exec::EvalCtx* context,
+    const SelectivityVector& rows,
+    FlatVector<StringView>& flatResult) {
+  // input is guaranteed to be in flat encoding when passed in.
+  auto inputArray = input.as<ArrayVector>();
+
+  auto elements = inputArray->elements();
+  auto elementsRows =
+      functions::toElementRows(elements->size(), rows, inputArray);
+  exec::LocalDecodedVector elementsHolder(context, *elements, elementsRows);
+  auto elementsDecoded = elementsHolder.get();
+
+  exec::LocalSelectivityVector elementsBaseRows(
+      context->execCtx(), elementsDecoded->base()->size());
+  elementsBaseRows->clearAll();
+  elementsRows.applyToSelected([&](auto row) {
+    elementsBaseRows->setValid(elementsDecoded->index(row), true);
+  });
+  elementsBaseRows->updateBounds();
```

Comment:
This stanza is repeated multiple times; can it be factored out?

Reply:
I extracted the computation of base rows into `translateRowsToBase()`. Thank you for this suggestion!